File: Leetcode/C++/142. Linked List Cycle II/solution.cpp
// https://leetcode.com/problems/linked-list-cycle-ii/
// Ref: https://leetcode.com/problems/linked-list-cycle-ii/solutions/44781/concise-o-n-solution-by-using-c-with-detailed-alogrithm-description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        if (head == NULL || head->next == NULL)
        {
            return NULL;
        }
        ListNode *slow = head;
        ListNode *fast = head;

        while (fast->next && fast->next->next)
        {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast)
            {
                while (slow != head)
                {
                    slow = slow->next;
                    head = head->next;
                }
                return head;
            }
        }
        return NULL;
    }
};

File: Leetcode/Python/142. Linked List Cycle II/solution.py
# https://leetcode.com/problems/linked-list-cycle-ii/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node_dict = {}
        node = head
        pos = -1
        while node and node.next:
            pos += 1
            if node in node_dict:
                return node
            print(node)
            node_dict[node] = pos
            node = node.next
    
        
        